Expensify/react-native-onyx
# Set default key states when clearing the cache

cc @marcaaron
### Details

When we clear Onyx and reset the cache set every value to `null` OR the default key state if one exists for that key. Remove the call to `initializeWithDefaultKeyStates` since it is no longer needed. These changes prevent updating Onyx subscribers with out of date information by sending only the defaultKeyStates. See the issue for more info and an example.

### Related Issues

https://github.com/Expensify/react-native-onyx/issues/125

### Automated Tests

[onyxClearNativeStorageTest.js](https://github.com/Expensify/react-native-onyx/blob/dcb7f0139f8b5371affdd3bb2f9ce84e4687f087/tests/unit/onyxClearNativeStorageTest.js) and [onyxClearWebStorageTest.js](https://github.com/Expensify/react-native-onyx/blob/068a98a9afd45a53114754cc96c6928ec9fccbb3/tests/unit/onyxClearWebStorageTest.js)

### Manual Tests
I had a lot of trouble adding an automated tests earlier so I also added manual tests.
1. Checkout the App branch for **Case 1**
2. `npm i` to install [this Onyx version](https://github.com/Expensify/react-native-onyx/tree/neil-test-clear) that has been modified for testing

For each case below:
1. Checkout the App branch
2. `npm run web`
3. Sign in to any account
4. Open the console
5. Sign out
6. Make sure there are no assertion errors in the console
7. For cases 1 and 2 only:
8. `command + f` in the console and search 'Onyx test'
9. Make sure that the cache clears, then data is set or merged, and then Onyx finishes clearing.

- [ ] **Case 1:** [neil-test-onyx-clear-merge](https://github.com/Expensify/App/tree/neil-test-onyx-clear-merge)

```
Onyx.clear();
Onyx.merge('someDefaultKey', 'someValue');

// Expect the value of the merge to replace the default key state
// Expect that storage matches what we have in the cache and that our call to 
// `Storage.clear()` did not clear the value of merge from storage after we called Onyx.clear()
```

- [ ] **Case 2:** [neil-test-onyx-clear-set](https://github.com/Expensify/App/tree/neil-test-onyx-clear-set)
```
Onyx.clear();
Onyx.set('someDefaultKey', 'someValue');

// Expect the value of the set to replace the default key state
// Expect that storage matches what we have in the cache and that our call 
// to `Storage.clear()` did not prevent any values from being set
```

- [ ] **Case 3:** [neil-test-onyx-set-clear](https://github.com/Expensify/App/tree/neil-test-onyx-set-clear)
```
Onyx.set('someDefaultKey', 'someValue');
Onyx.clear();

// Expect the value from the `set()` to be replaced with the default key state value
```

- [ ] **Case 4:** [neil-test-onyx-merge-clear](https://github.com/Expensify/App/tree/neil-test-onyx-merge-clear)
```
Onyx.merge('someDefaultKey', 'someValue');
Onyx.clear();

// Expect the value from the `merge()` to be replaced with the default key state value
```

### Linked PRs

https://github.com/Expensify/App/pull/8855
